cmd/aigw/translate.go

```
@@ -0,0 +1,261 @@
+)
+
+// translateFn is the function signature for the translate command for decoupling and testing.
+type translateFn func(cmd cmdTranslate, stdout, stderr io.Writer) error
+
+// translate implements the translateFn command. This function reads the input files, collects the AI Gateway custom resources,
+// translates them to Envoy Gateway and Kubernetes objects, and writes the translated objects to the output writer.
+func translate(cmd cmdTranslate, output, stderr io.Writer) error {
+	stderrLogger := slog.New(slog.NewTextHandler(stderr, &slog.HandlerOptions{}))
+	if !cmd.Debug {
+		stderrLogger = slog.New(slog.NewTextHandler(io.Discard, &slog.HandlerOptions{}))
+	}
+	var buf strings.Builder
+	for _, path := range cmd.Paths {
+		stderrLogger.Info("Reading file", "path", path)
+		content, err := os.ReadFile(path)
+		if err != nil {
+			return fmt.Errorf("error reading file %s: %w", path, err)
+		}
+		buf.Write(content)
+		buf.WriteRune('\n')
+	}
+
+	aigwRoutes, aigwBackends, backendSecurityPolicies, err := collectCustomResourceObjects(buf.String(), stderrLogger)
+	if err != nil {
+		return fmt.Errorf("error translating: %w", err)
+	}
+
+	err = translateCustomResourceObjects(aigwRoutes, aigwBackends, backendSecurityPolicies, output, stderrLogger)
+	if err != nil {
+		return fmt.Errorf("error emitting: %w", err)
+	}
+	return nil
+}
+
+// collectCustomResourceObjects reads the YAML input and collects the AI Gateway custom resources.
+func collectCustomResourceObjects(yamlInput string, logger *slog.Logger) (
+	aigwRoutes []*aigv1a1.AIGatewayRoute,
+	aigwBackends []*aigv1a1.AIServiceBackend,
+	backendSecurityPolicies []*aigv1a1.BackendSecurityPolicy,
+	err error,
+) {
+	decoder := yaml.NewYAMLOrJSONDecoder(bytes.NewReader([]byte(yamlInput)), 4096)
+	for {
+		var rawObj runtime.RawExtension
+		err = decoder.Decode(&rawObj)
+		if errors.Is(err, io.EOF) {
+			err = nil
+			return
+		} else if err != nil {
+			log.Fatalf("Error decoding YAML: %v", err)
```

`continue`?

---

i would fail instead of continuing as stderr logging is suppressed by default so users won't notice there's a wrong format file they are passing

---

oh wait anyways it should be log -> logger